gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java

```diff
@@ -542,17 +560,35 @@ public void start(Listener<RespT> responseListener, Metadata headers) {
             new SimpleForwardingClientCallListener<RespT>(responseListener) {
               @Override
               public void onClose(Status status, Metadata trailers) {
-                try {
-                  super.onClose(status, trailers);
-                } finally {
-                  entry.release();
+                if (wasClosed.compareAndSet(false, true)) {
+                  try {
+                    super.onClose(status, trailers);
+                  } finally {
+                    if (wasReleased.compareAndSet(false, true)) {
+                      entry.release();
+                    } else {
+                      LOG.log(
+                          Level.WARNING,
+                          "Entry was released before the call is closed. This may be due to an exception on start of the call.");
+                    }
+                  }
+                } else {
+                  LOG.log(
+                      Level.WARNING,
+                      "Call is being closed more than once. Please make sure that onClose() is not being manually called.");
                 }
```

Comment:
The nesting makes it hard to read, I think early returns would make it easier:

```java
if (wasClosed.compareAndSet(false, true)) {
   LOG.log(
                      Level.WARNING,
                      "Call is being closed more than once. Please make sure that onClose() is not being manually called.");
   return;
}

if (!wasReleased.compareAndSet(false, true)) {
    LOG.log(
                      Level.WARNING,
                      "Call is being closed more than once. Please make sure that onClose() is not being manually called.");
    return;
}
entry.release();
```

Reply:
Thanks for the suggestion! Yeah with try-finally being added, it's getting harder to read. I extracted the checking of `wasClosed` to earlier to remove one indent. Kept the checking of `wasRelease` in finally because I think it's logically easier to understand even though has more indents.